doc/developer/design/20230418_subscribe_output.md

```diff
@@ -0,0 +1,361 @@
+| `(k, v1, -1)` | `(k, delete, NULL)` | `(k, delete, v1, NULL)` |
+| `(k, v1, -1)`, `(k, v2, -1)` | `(k, delete, NULL)` | `(k, delete, v1, NULL)`, `(k, delete, v2, NULL)` |
+
+
+We don't have to handle unconsolidated data.
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+For each of the three modes we're only changing the results for a
+given timestamp. The sorting and changing the output format can be
+done in `ActiveSubscribe::process_response` which is guaranteed to
+have a complete `SubscribeResponse`'s batch of data. The returned
+batches [don't
+overlap](https://github.com/MaterializeInc/materialize/blob/e781222ea382aa6a7eee1b0e50e94218988f4cad/src/compute-client/src/protocol/response.rs#L106-L110)
+so we can operate on one batch at a time.
+
+`WITHIN TIMESTAMP ORDER BY`: in `process_response`, for each timestamp
+we'll sort the rows by the given `ORDER BY`. This `ORDER BY` can be as
+complicated as any `SELECT`. This adds adds some logic to compiling a
+`SUBSCRIBE` plan but we can reuse most of the existing planning
+machinery.
+
+`ENVELOPE UPSERT` and `ENVELOPE DEBEZIUM`: similar to `WITHIN
+TIMESTAMP ORDER BY` the transformation can be done locally in
+`process_response`: the set of retractions and additions uniquely
+determine the resulting upsert envelopes. The planning work is small:
+finding the indexes of the key columns. The changes to describe are a
+tad more interesting: we have to reorder columns and remove the
+`mz_diff` column.
+
+For `ENVELOPE UPSERT`, if all the updates to a key are retractions we
+emit a deletion for that key. Otherwise we emit an upsert for every
+added row.
+
+For `ENVELOPE DEBEZIUM`, for a given timestamp and key, sort the
+updates by `mz_diff`. Pair up as many retractions and insertions as
+possible and emit one debezium update for that pair. Any leftovers get
+their own debezium update that's either a sole insertion or
+deletion. For the normal case of a single live value for a key this
+algorithm does the right thing; when there are multiple values we
+favor updates over inserts or deletes. There's no guarantee on the
+order of the debezium updates or how values get paired together since
+from MZ's perspective all these updates happened in the same moment.
+
+# Rollout
+[rollout]: #rollout
+
+We'll put these options into production behind a launchdarkly flag
+that we can enable for clients if they're okay working without a
+backwards compatibility guarantee. That'll also let us gather feedback
+from clients using these features.
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+New testdrive tests that exercise these new output types.
+
+There isn't a performance concern because these output types only need
+to sort the data for a given timestamp.
+
+We'll add metrics to see how often these options are used in order to
```

Comment:
By metrics, do you mean segment events? Might be worth sketching out what we want to track here and running it by Graham as well. If you want separate that from the design doc that's also fine.

Reply:
I looked into it and I think segment events would work but i'll start with adding prometheus metrics to get a rough sense of usage first. added little note to doc